Feature/129158777 enable international sales order

j7nguyen opened the pull request:
## Description

Updates to gem that allow it to process international sales orders. Add comments to code and refactor to reduce redundant code and appease Rubocop.
## Changes
- [ ] update `SalesOrder`, `SalesOrderHeader`, `SalesOrderShipmentInformation`, `SalesOrderDetail`, and `SalesOrderLineItem` to accommodate additional xml fields needed for international shipments
- [ ] add `SalesOrderLineItemNameValue`, `SalesOrderHeaderNameValue`, and `SalesOrderLineItemSpecialMessage` as new elements in a sales order
- [ ] remove `customer_id` from `BaseElement` since it's not needed by all elements.
- [ ] replace sales order XSD file with recently provided file by Ingram Micro
- [ ] refactor other classes to clean up code and reduce redundancy

[#129158777]

Commit:
WIP Enabling international shipping

- updating Line Items in sales orders to include additional detail
- new element, line-name-value
- new element, header-name-value
- new element for special message
- new spec files

[#129158777]

Commit:
WIP Enabling international shipping

- updating Line Items in sales orders to include additional detail
- new element, line-name-value
- new element, header-name-value
- new element for special message
- new spec files

[#129158777]

Commit:
WIP adding comments and tests to incorporate new fields for international shipping

[#129158777]

Commit:
WIP adding comments and tests to incorporate new fields for international shipping

[#129158777]

Commit:
More updates to sales orders for international shipping

- SalesOrderLineItemNameValue updated to work in existing flow
- SalesOrderLineItem updated to utilize SalesOrderLineItemNameValue
- update specs

[#129158777]

Commit:
More updates to sales orders for international shipping

- SalesOrderLineItemNameValue updated to work in existing flow
- SalesOrderLineItem updated to utilize SalesOrderLineItemNameValue
- update specs

[#129158777]

Commit:
Updates for international shipping and refactoring

- add specs and fix SalesOrderLineItemSpecialMessage
- stop setting @element[:customer_id] on BaseElement (and everything inheriting from it)

[#129158777]

Commit:
Updates for international shipping and refactoring

- add specs and fix SalesOrderLineItemSpecialMessage
- stop setting @element[:customer_id] on BaseElement (and everything inheriting from it)

[#129158777]

Commit:
More refactoring and incorporating international shipping

- remove reference to customer_id where not needed
- add specs for method used to add special messages to sales order line items

[#129158777]

Commit:
More refactoring and incorporating international shipping

- remove reference to customer_id where not needed
- add specs for method used to add special messages to sales order line items

[#129158777]

Commit:
Merge with develop

Commit:
Consolidate files to just use one sales order schema for both domestic and international shipments

- remove unnecessary XSD files
- update transmission to reflect correct XSD file for sales order submissions
- remove unneeded SalesOrderInternational class

[#129158777]

Commit:
Renaming and refactoring

- spec file for transmission renamed from transmission.rb to transmission_spec.rb
- edits to appease Rubocop

[#129158777]

Commit:
Fix defaults in SalesOrderShipmentInformation

[#129158777]

Commit:
Refactoring to appease Rubocop and clean up code

[#129158777]

sent-hil commented on lib/ingram_micro/elements/credit_card_information.rb:
```diff
@@ -17,7 +17,7 @@ class IngramMicro::CreditCardInformation < IngramMicro::BaseElement
     avs_hold: nil,
     merchant_name: nil
-  }
+  }.freeze
```
👍

sent-hil commented on lib/ingram_micro/elements/outbound_purchase_order_information.rb:
```diff
@@ -6,14 +6,7 @@ class IngramMicro::OutboundPurchaseOrderInformation < IngramMicro::BaseElement
     comments: nil
-  }
-
-  def initialize(options = {})
```
why was this removed? irc i added this since we didn't need a `customer_id` here and having it there was problematic

j7nguyen replied:
I changed BaseElement, from which this inherits, to not automatically assign a `customer_id` to each element.

sent-hil commented on spec/ingram_micro/elements/sales_order_line_item_special_message_spec.rb:
```diff
@@ -0,0 +1,41 @@
+require 'pry'
+
+describe IngramMicro::SalesOrderLineItemSpecialMessage do
+
```
extra newline

sent-hil commented on lib/ingram_micro/elements/return_authorization_submission.rb:
```diff
@@ -1,67 +1,69 @@
-    if @element[:detail].element[:line_items].empty?
-      line_item = IngramMicro::ReturnAuthorizationLineItem.new
-      @element[:detail].element[:line_items] << line_item
+    def check_line_items
```
method does more than just `check`, should be renamed to convey that it does more

j7nguyen replied:
Removing this entirely as it is undesired behavior.

sent-hil commented on lib/ingram_micro/elements/sales_order_line_item.rb:
```diff
@@ -1,37 +1,79 @@
+      # Expect that line_name_value will be an array of arrays, each of which
+      # contains a name and value that will translate to line-attribute-name and
+      # line-attribute-value.
+      unless element[:line_name_value].nil?
```
couldn't this be written as `if element[:line_name_value]`

sent-hil commented on lib/ingram_micro/elements/sales_order_line_item.rb:
```diff
@@ -1,37 +1,79 @@
+      # line-attribute-value.
+      unless element[:line_name_value].nil?
+        element[:line_name_value].each do |pair|
+          attr_name, attr_value = pair
```
did you do this on purpose? `attr_value` is assigned to nil

j7nguyen replied:
It's being assigned to the second element in pair, which is an array. Perhaps this isn't obvious and I should instead do `attr_name = pair[0]` and `attr_value = pair[1]`

sent-hil commented on lib/ingram_micro/elements/sales_order_line_item.rb:
```diff
@@ -1,37 +1,79 @@
+    # initialization to make this work.
+    def add_special_message(builder)
+      message = element[:special_message]
+      message.build(builder) unless message.nil?
```
`if message`

sent-hil commented on lib/ingram_micro/base_element.rb:
```diff
@@ -1,18 +1,22 @@
       value = (options[field] ? options[field] : default)
       @element[field] = value
     end
-    @element[:customer_id] ||= IngramMicro.configuration.customer_id
```
wouldn't removing this be problematic? are you sure this doesn't break the xml we send to IM

j7nguyen replied:
I think this is better, as we don't need `customer_id` set on every element. This way, we are more explicit about setting it in the necessary places.

sent-hil commented on lib/ingram_micro/elements/return_authorization_submission.rb:
```diff
@@ -1,67 +1,69 @@
-  def defaults
-    DEFAULTS
-  end
+    def defaults
```
is there a reason you're redefining this in each class? if you only put it in base class and sub class it, does it look up the right `DEFAULT` ?

sent-hil replied:
@vietjew missed ^ comment

sent-hil commented on lib/ingram_micro/elements/sales_order_line_item_name_value.rb:
```diff
@@ -0,0 +1,33 @@
+module IngramMicro
+  class SalesOrderLineItemNameValue < BaseElement
+    ATTRIBUTE_NAMES = [
```
what's the difference between this and `HEADER_ATTRIBUTE_NAMES`

j7nguyen replied:
They're the acceptable values for a different element, which perhaps should be implied more in the constant name.

sent-hil commented on lib/ingram_micro/elements/sales_order_submission.rb:
```diff
@@ -1,67 +1,68 @@
-    if @element[:line_items].empty?
-      line_item = IngramMicro::SalesOrderLineItem.new
-      @element[:line_items] << line_item
+    def check_line_items
```
method does more than just check, should be renamed to convey that it does more

j7nguyen replied:
Removing this altogether so that if no line items are passed in, the xml won't pass the schema.

sent-hil commented on lib/ingram_micro/elements/shipment_status.rb:
```diff
@@ -1,44 +1,47 @@
-    if @element[:line_items].empty?
-      ss_line_item = IngramMicro::ShipmentStatusLineItem.new
-      @element[:line_items] << ss_line_item
+    def check_line_items
```
method does more than just check, should be renamed to convey that it does more

j7nguyen replied:
I had just rewritten this to appease rubocop, but now that I look more closely I don't think this method offers desired behavior. It's adding an empty line item if there are none, whereas it should raise an error.

sent-hil commented on lib/ingram_micro/elements/shipment_status.rb:
```diff
@@ -1,44 +1,47 @@
-      raise IngramMicro::InvalidType.new('customer_id must be a number') unless integer?(@element[:customer_id])
-    true
+    def valid?
+      unless integer?(@element[:customer_id])
```
feels weird having a private method called `integer?` does customer_id have to be a Integer but in a string?

sent-hil commented on lib/ingram_micro/transmission.rb:
```diff
@@ -8,19 +8,24 @@ class IngramMicro::Transmission
+  # schema_valid? compares an xml document (the result of self.xml_builder) to
+  # the appropriate xsd (XML Schema Definition).
+  # NOTE: the order of the fields matters when checking against the xsd files.
+  #   So, if in the order-header of a sales order, if customer_channel_type
```
include `, but if customer_channel_type appears at end of xsd file, it will fail...`

sent-hil commented on lib/ingram_micro/transmission.rb:
```diff
@@ -44,12 +49,12 @@ def add_transaction_info(builder)
   end
 
   def send_request
     client = IngramMicro::Client.new
```
not need to store this in a new variable, can be inlined

j7nguyen replied:
Good call, fixed.

sent-hil commented on spec/ingram_micro/elements/sales_order_line_item_spec.rb:
```diff
@@ -0,0 +1,98 @@
+require 'pry'
+
+describe IngramMicro::SalesOrderLineItem do
+  # let(:populated_sales_order) { Fabricate.build(:sales_order) }
```
remove this and line below

sent-hil commented on spec/ingram_micro/elements/sales_order_line_item_spec.rb:
```diff
@@ -0,0 +1,98 @@
+  # let(:populated_sales_order) { Fabricate.build(:sales_order) }
+  # let(:line_item_1) { populated_sales_order.detail.element[:line_items].first }
+  let(:line_item) { Fabricate.build(:sales_order_line_item) }
+  describe '#line_no and #line_no=' do
```
space between let and describe

sent-hil commented on spec/ingram_micro/elements/sales_order_line_item_spec.rb:
```diff
@@ -0,0 +1,98 @@
+
+  describe '#add_line_name_value' do
+    let(:line_attr_name) { 'international-declared-value' }
+    let(:line_attr_value) { '1999' }
```
space between let and context

sent-hil commented on spec/ingram_micro/elements/sales_order_line_item_spec.rb:
```diff
@@ -0,0 +1,98 @@
+      expect(builder.to_xml).to include('<special-message>')
+      expect(builder.to_xml).to include('<engraving-font>hellavetica')
+    end
+
```
unnecessary new line

sent-hil commented on spec/ingram_micro/elements/sales_order_line_item_spec.rb:
```diff
@@ -0,0 +1,98 @@
+        end
+
+        it 'properly creates xml' do
+          li = described_class.new(product_name: "killerizer", line_name_value: [["international-license-value",'300.50']])
```
i personally don't like described_class because each time i've to scroll up and see what the class is...not need to change it, just a fyi

j7nguyen replied:
I see what you mean, but with these really long class names I find described_class helpful.

sent-hil commented:
@vietjew done reviewing. If any of the comments don't make sense, we can pair tuesday. Please assign it back to me when done fixing/replying to the comments.

Commit:
Remove unnecessary methods and clean up code

- remove `check_line_items` from multiple classes to avoid passing schema validation when no line items are passed into elements that require them
- fix comments to be more helpful
- fix conditional statements and inline to clean up code

[#129158777]
